Bt shell improvements

joerchan opened the pull request:
Add command to print connection info.
Handle registering authentication callback error code.

joerchan committed:
Bluetooth: shell: Add info command to print connection information

Add command to print connection info object.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: shell: Handle error code of bt_conn_auth_cb_register

Fix issue with registering authentication callback handlers failing
without notifying the user of the shell.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

carlescufi commented:
@joerchan thanks! can you post a snippet of the output of the `info` command?

carlescufi reviewed (approved):
I have no objections to the code, but would it also be a good idea to provide a mechanism to print info for all existing connections?

Vudentz commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -787,6 +787,101 @@ static int cmd_select(const struct shell *shell, size_t argc, char *argv[])
+
+	switch (argc) {
+	case 1:
+		if (default_conn) {
```
Perhaps by default we should list all connections not just the default.

joerchan replied:
How about adding "all" as argument to print all connections?

joerchan replied:
I would need #17426 to iterate over all connections.

Vudentz commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -787,6 +787,101 @@ static int cmd_select(const struct shell *shell, size_t argc, char *argv[])
+
+	if (info.type == BT_CONN_TYPE_LE) {
+		bt_addr_le_to_str(info.le.src, addr_str, sizeof(addr_str));
+		shell_print(ctx_shell, "Initiator  address: %s", addr_str);
```
Id use Source and Destination terminology for addresses since we may not always be the central as this seems to assume.

carlescufi replied:
I would use the GAP terms: Central and Peripheral, when printing to the shell

Vudentz replied:
> I would use the GAP terms: Central and Peripheral, when printing to the shell

Afaik they are not role depend, so src always represents the local source address and dst is always the remote destination address so we would have to check the role of the connection to do what you are suggesting which imo complicates more necessary.

joerchan replied:
Maybe it should have been "initiator" and "responder" I'm not really sure which names are best here, but that correlates with "init_addr" and "resp_addr"
I'd like to avoid confusion here with the "dst" addr which is returned by `bt_conn_get_dst`, this is the address that should be used when re-connecting for example. It represents the resolved identity address of the peer, or the normal address if privacy is not used or not resolved yet.
The init_addr is the address that was sent on-air in the connection creation. It is used for example by SMP and will not change during the connection.

Vudentz replied:
Under what situation does an application need to know the init_addr and resp_addr? To me it should like this should be handled by stack, in fact we already have code handling similar situation in bt_conn_addr_le_cmp.

Vudentz replied:
Id stick with Source/Local and Destination/Remote addresses then since get info is doing the right thing when assigning the addresses by checking the connection role, the only thing that perhaps is missing is to use identity addresses if there is one so that would be consistent with get_dst, though the same can be said for local address as well where the address depend on the id used for connecting.

joerchan replied:
> Id stick with Source/Local and Destination/Remote addresses then since get info is doing the right thing when assigning the addresses by checking the connection role

You are right, messed that up, 

>  the only thing that perhaps is missing is to use identity addresses if there is one so that would be consistent with get_dst

In that case we should always return dst.
Not sure what is the best thing to do here.

1. Info was meant to provide the source and destination, in which case we should have returned `&conn->le.dst` and `&bt_dev.id_addr[conn->id]`
2. Info is correct, but documentation is not clear that it is the device addresses used on-air, improve documentation?. Also there is no way to get out the source address, maybe add `bt_conn_get_src`?
3. Add all addresses in get_info? src/dst and resp/init

Ideas?

Vudentz replied:
3. might be the way to go, it seems we have the role already exposed so the application can make sense of responder and initiator addresses and the shell can be a good source to how to use them properly. As for the bt_conn_get_src that might be a good idea as well, either that or bt_conn_get_id.

joerchan commented:
> @joerchan thanks! can you post a snippet of the output of the `info` command?

```
Peer address: 54:9c:a7:fb:82:77 (random)
Type: LE, Role: slave, Id 0
Initiator  address: 55:94:4f:05:93:72 (random)
Advertiser address: 54:9c:a7:fb:82:77 (random)
interval: 0x0028 (50.00 ms)
latency: 0x0000 (0.00 ms)
timeout: 0x002a (420 ms)
```

Vudentz commented:
> > @joerchan thanks! can you post a snippet of the output of the `info` command?
> 
> ```
> Peer address: 54:9c:a7:fb:82:77 (random)
> Type: LE, Role: slave, Id 0
> Initiator  address: 55:94:4f:05:93:72 (random)
> Advertiser address: 54:9c:a7:fb:82:77 (random)
> interval: 0x0028 (50.00 ms)
> latency: 0x0000 (0.00 ms)
> timeout: 0x002a (420 ms)
> ```
> 
> So the info struct is really carrying the init_addr and resp_addr as src and dest? That is very misleading imo, we should probably rename these to avoid confusion in the future.
> 
> Btw, I would have assumed that by being slave the peer address would be initiator or is the problem here is that bt_conn_get_info swaps src/dst depending on the role, so we got src = resp_addr and dst = init_addr, no wonder it is hard to understand the meaning of these.

Ouch looks like I managed to edit your post @joerchan instead of quote it, my bad.

joerchan commented:
> So the info struct is really carrying the init_addr and resp_addr as src and dest? That is very misleading imo, we should probably rename these to avoid confusion in the future.

Yes, that is the confusing. Also since the function `bt_conn_get_dst` exists, it is unclear what is the correct way. The functions seems to have been added at the same time, and also before privacy. 
Maybe this was not the intention of the APIs, and we should clean it up?  There has been problems in the stack as well, see #17981.

joerchan committed:
Bluetooth: Host: Add identity addresses to conn info object

Use the src and dst naming to refer to the identity addresses of the
connection. Keep the device addresses used during connections but rename
them to local and remote instead.
Update documentation to be more descriptive.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: Shell: Use on-air addresses when checking for OOB data

Authentication using OOB data uses the on-air device addresses used
during connection setup. So we need to check against the on-air
addresses in the info object to see if we have correct OOB data.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
Update info object to carry source identity and destination identity addresses in src/dst. added local/remote for on-air addresses.
Output looks like this: (after pairing)

```
Type: LE, Role: master, Id: 0
Remote address: c1:f6:39:6f:90:c5 (random) (identity)
Local address: f1:c1:d9:aa:49:e2 (random) (identity)
Remote on-air address: 5b:ce:8b:94:62:69 (random) (resolvable)
Local on-air address: 54:08:d1:10:57:fa (random) (resolvable)
interval: 0x0028 (50.00 ms)
latency: 0x0000 (0.00 ms)
timeout: 0x002a (420 ms)
```

Vudentz reviewed (approved):
LGTM
